src/SimpleCollection/AbstractCollection.php

```diff
@@ -344,39 +340,33 @@ public function getKeys()
     }
 
     /**
-     * Filters the current values and return a new collection
-     *
+     * Produces an immutable version of the list and filters for values
      * @param \Closure $cClosure
-     *
      * @return AbstractCollection
      */
     public function filter(\Closure $cClosure)
     {
-        $sClassName      = get_class($this);
-        $aFilteredValues = array();
-        foreach ($this->values as $sKey => $mValue) {
-            if (true === $cClosure($mValue, $sKey)) {
-                $aFilteredValues[$sKey] = $mValue;
+        $modifiedCollection = clone $this;
+        foreach ($modifiedCollection->values as $sKey => $mValue) {
+            if (false === call_user_func($cClosure, $mValue, $sKey)) {
+                unset($modifiedCollection->values[$sKey]);
             }
         }
-
-        return new $sClassName($aFilteredValues);
+        return $modifiedCollection;
     }
 
     /**
-     * Use a function on all values of the collection, and set the result as new values for the key
-     *
+     * Method too apply a closure to all values available. Will return an immutable state.
      * @param \Closure $cClosure
-     *
-     * @return $this
+     * @return AbstractCollection
      */
     public function forAll(\Closure $cClosure)
```

Comment:
Imho this is no case of a immutable state, because you want to change to current collection.
Btw there is a major break too